### PR TITLE
Instant Search: Add support for filters in Redux store

### PR DESCRIPTION
**Pull request by jsnmoon**
Follow-up for #18118 and #18119.

#### Changes proposed in this Pull Request:
* Add support for handling filters in Redux store. Does not integrate with the Instant Search component tree.
* Fixes ESLint warnings for the filters library.
* Adds a new, tested function to the filters library.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

The Redux store is not yet connected up to the Instant Search front end, so this PR should not affect the existing search functionality.

* Review the code and ensure it looks reasonable.
* Run `yarn test-search` and ensure all tests pass.

#### Proposed changelog entry for your changes:
* None.

### CODE REVIEW

File: modules/search/instant-search/lib/filters.js
Comment (jsnmoon) [inline modules/search/instant-search/lib/filters.js:8]: Changes to this file can be ignored; these changes are identical to #18119.

File: modules/search/instant-search/lib/query-string.js
Comment (jsnmoon) [inline modules/search/instant-search/lib/query-string.js:27]: In hindsight, these changes should have landed in #18118. However, this isn't a big concern given that the `effects.js` module doesn't get loaded until it's integrated with the component tree.

File: modules/search/instant-search/lib/filters.js
Comment (bluefuton) [inline modules/search/instant-search/lib/filters.js:74]: Can we add a test in this PR? 😇

Comment (jsnmoon) [reply modules/search/instant-search/lib/filters.js:74]: 👍 I'll add the test after landing and rebasing off of #18119.

Keeping this comment open until then.

Comment (jsnmoon) [reply modules/search/instant-search/lib/filters.js:111]: Tests have now been added; this is ready for another review.

**Comment by jsnmoon** [comment]
This is ready for another round of reviews.

**Comment by jsnmoon** [comment]
Rebased to fix merge conflicts.

**Review by bluefuton — APPROVED** [review]
Nice work on the tests + JSdoc! 💯

**Review by jeherve — APPROVED** [review]

**Comment by jsnmoon** [comment]
r219373-wpcom

**Pull request merged**
